pkg/sql/span/span_builder.go

```diff
@@ -75,73 +76,98 @@ func (s *Builder) SpanFromEncDatums(
+	isDesc := s.keyAndPrefixCols[prefixLen].Direction == catpb.IndexColumn_DESC
 	if isDesc {
-		startDatum, endDatum = endDatum, startDatum
+		startBound, endBound = endBound, startBound
 		startInclusive, endInclusive = endInclusive, startInclusive
 	}
 
+	if startBound != nil {
+		if err = startBound.EnsureDecoded(rangeColTyp, &s.alloc); err != nil {
+			return roachpb.Span{}, false, false, err
+		}
+		if startBound.IsNull() {
+			// Inequalities are null-rejecting.
+			return roachpb.Span{}, true, true, nil
+		}
+		if !startInclusive {
+			if (isDesc && startBound.Datum.IsMin(s.evalCtx)) ||
+				(!isDesc && startBound.Datum.IsMax(s.evalCtx)) {
+				// There are no values that satisfy the start bound.
+				return roachpb.Span{}, false, true, nil
+			}
+		}
+	}
+	if endBound != nil {
+		if err = endBound.EnsureDecoded(rangeColTyp, &s.alloc); err != nil {
+			return roachpb.Span{}, false, false, err
+		}
+		if endBound.IsNull() {
+			// Inequalities are null-rejecting.
+			return roachpb.Span{}, true, true, nil
+		}
+	}
+
 	makeKeyFromRow := func(r rowenc.EncDatumRow) (_ roachpb.Key, containsNull bool, _ error) {
 		return rowenc.MakeKeyFromEncDatums(r, s.keyAndPrefixCols, &s.alloc, s.KeyPrefix)
 	}
 
 	var startKey, endKey roachpb.Key
 	var startContainsNull, endContainsNull bool
-	if startDatum != nil {
-		if !startInclusive {
-			// Since the spans are defined such that the start boundary is
-			// inclusive, yet our inequality condition wants to exclude it,
-			// we'll "advance" the start datum by one.
+	if startBound != nil {
+		startDatum := startBound.Datum
+		if !startInclusive && isDesc {
+			// We need the datum that sorts immediately before startDatum in order
+			// to make the start boundary inclusive. The optimizer should have
+			// filtered cases where this is not possible. If the index column is ASC,
+			// we can directly increment the key below instead of the datum.
 			var ok bool
-			if isDesc {
-				startDatum, ok = startDatum.Prev(s.evalCtx)
-			} else {
-				startDatum, ok = startDatum.Next(s.evalCtx)
-			}
+			startDatum, ok = startDatum.Prev(s.evalCtx)
 			if !ok {
-				// The optimizer should have checked that the datum can be
-				// "advanced" when construing the range-based lookup join, so
-				// this is unexpected.
-				return roachpb.Span{}, false, errors.AssertionFailedf(
-					"couldn't get a Next or Prev value for %s", startDatum,
+				return roachpb.Span{}, false, false, errors.AssertionFailedf(
+					"couldn't get a Prev value for %s", startBound.Datum,
 				)
 			}
 		}
-		values[prefixLen-1] = rowenc.EncDatum{Datum: startDatum}
-		startKey, startContainsNull, err = makeKeyFromRow(values[:prefixLen])
+		values[prefixLen] = rowenc.EncDatum{Datum: startDatum}
+		startKey, startContainsNull, err = makeKeyFromRow(values[:prefixLen+1])
+		if !startInclusive && !isDesc {
+			// We can make the key inclusive by incrementing to the key that sorts
+			// immediately after it.
+			startKey = startKey.PrefixEnd()
```

Comment:
I'd like to draw attention to this bit here - does it work to call `PrefixEnd` in order to make the start key inclusive? Calling `Key.Next` didn't seem to properly filter out the previous value when I tried it.